README.md
# metastats-python-sdk



changelog.md
3.2.2
  - update package info

3.2.1
  - fix faq url

3.2.0
  - added riskInBalancePercent metric to trades and open trades
  - added riskInPips metric to trades and open trades

3.1.2
  - fixed dependency vulnerabilities

3.1.1
  - added references to MT manager api and risk management api

3.1.0
  - added region support
  - breaking change: changed default value for update_history parameter get_account_trades method
  - improved readme and examples

2.0.0
  - added getAccountTrades method
  - added getAccountOpenTrades method
  - added freeMargin metric
  - added margin metric
  - added marginLevel metric
  - added cagr metric
  - added mar metric

1.0.4
  - updated dependencies

1.0.3
  - documentation clarifications for some metrics

1.0.0
  - Implemented MetaStats SDK


lib/clients/httpClient_test.py
from .httpClient import HttpClient
import re
import pytest
import respx
from datetime import datetime
import json
from httpx import Response
from ..models import format_date
httpClient: HttpClient = None
test_url = 'http://example.com'
opts = {}


@pytest.fixture(autouse=True)
async def run_around_tests():
    global httpClient
    httpClient = HttpClient(retry_opts={'minDelayInSeconds': 0.05, 'maxDelayInSeconds': 0.2})
    global opts
    opts = {'url': test_url}
    yield


class TestHttpClient:
    @pytest.mark.asyncio
    async def test_load(self):
        """Should load HTML page from example.com"""
        response = await httpClient.request(opts)
        text = response.text
        assert re.search('doctype html', text)

    @pytest.mark.asyncio
    async def test_not_found(self):
        """Should return NotFound exception if server returns 404"""
        opts = {
            'url': f'{test_url}/not-found'
        }
        try:
            await httpClient.request(opts)
            raise Exception('NotFoundException is expected')
        except Exception as err:
            assert err.__class__.__name__ == 'NotFoundException'

    @pytest.mark.asyncio
    async def test_timeout(self):
        """Should return ConnectTimeout exception if request is timed out"""
        httpClient = HttpClient(0.001, {'retries': 2, 'minDelayInSeconds': 0.05, 'maxDelayInSeconds': 0.2})
        try:
            await httpClient.request(opts)
            raise Exception('ConnectTimeout is expected')
        except Exception as err:
            assert err.__class__.__name__ == 'ConnectTimeout'

    @respx.mock
    @pytest.mark.asyncio
    async def test_retry_on_api_exception(self):
        """Should retry request on fail with ApiException exception."""
        respx.get(test_url).mock(side_effect=[Response(502), Response(502),
                                              Response(200, content=json.dumps('response'))])
        response = await httpClient.request(opts)
        assert response == 'response'
        assert respx.get(test_url).call_count == 3

    @respx.mock
    @pytest.mark.asyncio
    async def test_retry_on_internal_exception(self):
        """Should retry request on fail with InternalException exception."""
        respx.get(test_url).mock(side_effect=[Response(500), Response(500),
                                              Response(200, content=json.dumps('response'))])
        response = await httpClient.request(opts)
        assert response == 'response'
        assert respx.get(test_url).call_count == 3

    @respx.mock
    @pytest.mark.asyncio
    async def test_return_error_on_retry_limit_exceeded(self):
        """Should return error if retry limit exceeded."""
        respx.get(test_url).mock(side_effect=Response(502))
        httpClient = HttpClient(60, {'retries': 2})
        try:
            await httpClient.request(opts)
            raise Exception('ApiException is expected')
        except Exception as err:
            assert err.__class__.__name__ == 'ApiException'
        assert respx.get(test_url).call_count == 3

    @respx.mock
    @pytest.mark.asyncio
    async def test_not_retry_if_exception_not_internal_exception_or_api_exception(self):
        """Should not retry if exception is neither InternalException nor ApiException."""
        error = {
            'id': 1,
            'error': 'error',
            'message': 'test',
            'details': [{'parameter': 'password', 'value': 'wrong', 'message': 'Invalid value'}]
        }
        respx.get(test_url).mock(side_effect=[Response(400, json=error), Response(400, json=error), Response(204)])
        try:
            await httpClient.request(opts)
            raise Exception('ValidationException is expected')
        except Exception as err:
            assert err.__class__.__name__ == 'ValidationException'
            assert err.__str__() == 'test, check error.details for more information'
            assert err.details == error['details']
        assert respx.get(test_url).call_count == 1

    def get_too_many_requests_error(self, sec):
        date = datetime.now().timestamp()
        date += sec
        recommended_retry_time = format_date(datetime.fromtimestamp(date))
        return Response(429, content=json.dumps({"metadata": {"recommendedRetryTime": recommended_retry_time}}))

    @respx.mock
    @pytest.mark.asyncio
    async def test_retry_after_wait_on_too_many_requests_error(self):
        """Should retry request after waiting on fail with TooManyRequestsException error."""
        respx.get(test_url).mock(side_effect=[self.get_too_many_requests_error(2),
                                              self.get_too_many_requests_error(3),
                                              Response(200, content=json.dumps('response'))])
        response = await httpClient.request(opts)
        assert response == 'response'
        assert respx.get(test_url).call_count == 3

    @respx.mock
    @pytest.mark.asyncio
    async def test_return_error_if_recommended_time_too_long(self):
        """Should return error if recommended retry time is too long."""
        respx.get(test_url).mock(side_effect=[self.get_too_many_requests_error(2),
                                              self.get_too_many_requests_error(300),
                                              Response(200, content=json.dumps('response'))])
        try:
            await httpClient.request(opts)
            raise Exception('TooManyRequestsException is expected')
        except Exception as err:
            assert err.__class__.__name__ == 'TooManyRequestsException'
        assert respx.get(test_url).call_count == 2

    @respx.mock
    @pytest.mark.asyncio
    async def test_not_count_retrying_too_many_requests_exception(self):
        """Should not count retrying TooManyRequestsException error."""
        respx.get(test_url).mock(side_effect=[self.get_too_many_requests_error(2), Response(502),
                                              Response(200, content=json.dumps('response'))])
        httpClient = HttpClient(60, {'retries': 1})
        response = await httpClient.request(opts)
        assert response == 'response'
        assert respx.get(test_url).call_count == 3

    @respx.mock
    @pytest.mark.asyncio
    async def test_wait_for_retry_after_header(self):
        """Should wait for the retry-after header time before retrying."""
        httpClient = HttpClient()
        respx.get(test_url).mock(side_effect=[
            Response(202, headers={'retry-after': '3'}),
            Response(202, headers={'retry-after': '3'}),
            Response(200, content=json.dumps('response'))])
        response = await httpClient.request(opts)
        assert response == 'response'
        assert respx.get(test_url).call_count == 3

    @respx.mock
    @pytest.mark.asyncio
    async def test_return_error_if_retry_after_time_too_long(self):
        """Should return TimeoutException error if retry-after header time is too long."""
        httpClient = HttpClient(60, {'maxDelayInSeconds': 3})
        respx.get(test_url).mock(side_effect=[
            Response(202, headers={'retry-after': '30'}),
            Response(202, headers={'retry-after': '30'}),
            Response(200, content=json.dumps('response'))])
        try:
            await httpClient.request(opts)
            raise Exception('TimeoutException is expected')
        except Exception as err:
            assert err.__class__.__name__ == 'TimeoutException'
            assert err.args[0] == 'Timed out waiting for the end of the process of calculating metrics'
        assert respx.get(test_url).call_count == 1

    @respx.mock
    @pytest.mark.asyncio
    async def test_return_error_if_timed_out_to_retry(self):
        """Should return TimeoutException error if timed out to retry."""
        respx.get(test_url).mock(side_effect=Response(202, headers={'retry-after': '0.9'}))
        httpClient = HttpClient(60, {'maxDelayInSeconds': 2, 'retries': 3})
        try:
            await httpClient.request(opts)
            raise Exception('TimeoutException is expected')
        except Exception as err:
            assert err.__class__.__name__ == 'TimeoutException'
            assert err.args[0] == 'Timed out waiting for the end of the process of calculating metrics'
        assert respx.get(test_url).call_count == 6


lib/clients/metaStats_client_test.py
import pytest
import respx
from httpx import Response
from .httpClient import HttpClient
from .metaStats_client import MetaStatsClient

METASTATS_API_URL = 'https://metastats-api-v1.agiliumtrade.agiliumtrade.ai'
token = 'header.payload.sign'
expected = {
    'trades': 10,
    'equity': 10102.5,
    'balance': 10105,
    'profit': 104,
    'deposits': 10001
}
account_id = '1234567'
httpClient = HttpClient()
metastats_client = MetaStatsClient(httpClient, token)


class TestGetMetrics:
    @respx.mock
    @pytest.mark.asyncio
    async def test_retrieve_metrics(self):
        """Should retrieve account metrics from API."""
        rsps = respx.get(f'{METASTATS_API_URL}/users/current/accounts/{account_id}/metrics?'
                         f'includeOpenPositions=false') \
            .mock(return_value=Response(200, json={'metrics': expected}))
        profiles = await metastats_client.get_metrics(account_id)
        assert rsps.calls[0].request.url == \
            f'{METASTATS_API_URL}/users/current/accounts/{account_id}/metrics?includeOpenPositions=false'
        assert rsps.calls[0].request.method == 'GET'
        assert rsps.calls[0].request.headers['auth-token'] == token
        assert profiles == expected

    @respx.mock
    @pytest.mark.asyncio
    async def test_retrieve_metrics_with_open_positions(self):
        """Should retrieve account metrics with included open positions from API."""
        rsps = respx.get(f'{METASTATS_API_URL}/users/current/accounts/{account_id}/metrics?'
                         f'includeOpenPositions=true') \
            .mock(return_value=Response(200, json={'metrics': expected}))
        profiles = await metastats_client.get_metrics(account_id, True)
        assert rsps.calls[0].request.url == \
            f'{METASTATS_API_URL}/users/current/accounts/{account_id}/metrics?includeOpenPositions=true'
        assert rsps.calls[0].request.method == 'GET'
        assert rsps.calls[0].request.headers['auth-token'] == token
        assert profiles == expected


start_time = '2020-01-01 00:00:00.000'
end_time = '2021-01-01 00:00:00.000'


class TestGetTrades:

    @respx.mock
    @pytest.mark.asyncio
    async def test_retrieve_account_trades(self):
        """Should retrieve account trades from API."""
        rsps = respx.get(f'{METASTATS_API_URL}/users/current/accounts/{account_id}/historical-trades/'
                         f'{start_time}/{end_time}') \
            .mock(return_value=Response(200, json={'trades': expected}))
        trades = await metastats_client.get_account_trades(account_id, start_time, end_time)
        assert rsps.calls[0].request.url == \
            f'{METASTATS_API_URL}/users/current/accounts/{account_id}/historical-trades/2020-01-01%2000:00:00.000' \
            '/2021-01-01%2000:00:00.000?updateHistory=false&limit=1000&offset=0'
        assert rsps.calls[0].request.method == 'GET'
        assert rsps.calls[0].request.headers['auth-token'] == token
        assert trades == expected


class TestGetOpenTrades:

    @respx.mock
    @pytest.mark.asyncio
    async def test_retrieve_account_open_trades(self):
        """Should retrieve account open trades from API."""
        rsps = respx.get(f'{METASTATS_API_URL}/users/current/accounts/{account_id}/open-trades') \
            .mock(return_value=Response(200, json={'openTrades': expected}))
        open_trades = await metastats_client.get_account_open_trades(account_id)
        assert rsps.calls[0].request.url == \
            f'{METASTATS_API_URL}/users/current/accounts/{account_id}/open-trades'
        assert rsps.calls[0].request.method == 'GET'
        assert rsps.calls[0].request.headers['auth-token'] == token
        assert open_trades == expected


lib/metaStats.py
from .clients.httpClient import HttpClient, RetryOpts
from .clients.metaStats_client import MetaStatsClient
from typing_extensions import TypedDict
from typing import Optional
import traceback


class ConnectionOptions(TypedDict):
    """Connection options."""
    requestTimeout: Optional[float]
    """Request timeout in seconds, default is 60."""
    domain: Optional[str]
    """Request domain, default 'agiliumtrade.agiliumtrade.ai.'"""
    retryOpts: Optional[RetryOpts]
    """Retry options."""


class MetaStats:
    """MetaStats API SDK."""

    def __init__(self, token: str, opts: ConnectionOptions = None):
        """Inits MetaStats class instance.

        Args:
            token: Authorization token.
            opts: Connection options.
        """
        opts: ConnectionOptions = opts or {}
        http_client = HttpClient(opts['requestTimeout'] if 'requestTimeout' in opts else None,
                                 opts['retryOpts'] if 'retryOpts' in opts else None)
        self._metaStatsClient = MetaStatsClient(http_client, token, opts['domain'] if 'domain' in opts else None)

    @property
    def get_metrics(self):
        """Returns the get_metrics MetaStatsClient method bound to the MetaStatsClient instance.

        Returns:
            get_metrics MetaStatsClient method.
        """
        return self._metaStatsClient.get_metrics

    @property
    def get_account_trades(self):
        """Returns the get_account_trades MetaStatsClient method bound to the MetaStatsClient instance.

        Returns:
            get_account_trades MetaStatsClient method.
        """
        return self._metaStatsClient.get_account_trades

    @property
    def get_account_open_trades(self):
        """Returns the get_account_open_trades MetaStatsClient method bound to the MetaStatsClient instance.

        Returns:
            get_account_open_trades MetaStatsClient method.
        """
        return self._metaStatsClient.get_account_open_trades

    @staticmethod
    def format_error(err: Exception):
        """Formats and outputs metaApi errors with additional information.

        Args:
            err: Exception to process.
        """
        error = {'name': err.__class__.__name__, 'message': err.args[0]}
        if hasattr(err, 'status_code'):
            error['status_code'] = err.status_code
        if err.__class__.__name__ == 'ValidationException':
            error['details'] = err.details
        if err.__class__.__name__ == 'TooManyRequestsException':
            error['metadata'] = err.metadata
        error['trace'] = traceback.format_exc()
        return error


lib/clients/errorHandler.py
from typing import Tuple, List
from typing_extensions import TypedDict


class ApiException(Exception):
    """Base class for API exceptions. Contains indication of HTTP status.

    Attributes:
        status_code: HTTP status code
    """

    def __init__(self, message: str, status: int):
        """Inits ApiException

        Args:
            message: Exception message.
            status: HTTP status.
        """
        super().__init__(message)
        self.status_code = status

    @property
    def code(self) -> str:
        """Returns exception code used for i18n

        Returns:
            Exception code.
        """
        return self._code

    @code.setter
    def code(self, code: str):
        """Sets error code, used for i18n

        Args:
            code: Error code for i18n.
        """
        self._code = code

    @property
    def arguments(self) -> Tuple:
        """Returns message arguments for i18n

        Returns:
            Message arguments for i18n.
        """
        return self.args

    @arguments.setter
    def arguments(self, args: Tuple):
        """Set message arguments for i18n

        Args:
            args: Arguments for i18n.
        """
        self.args = args


class NotFoundException(ApiException):
    """Throwing this exception results in 404 (Not Found) HTTP response code."""

    def __init__(self, message: str):
        """Inits not found exception.

        Args:
            message: Exception message.
        """
        super().__init__(message, 404)


class ForbiddenException(ApiException):
    """Throwing this exception results in 403 (Forbidden) HTTP response code."""

    def __init__(self, message: str):
        """Inits forbidden exception.

        Args:
            message: Exception message.
        """
        super().__init__(message, 403)


class UnauthorizedException(ApiException):
    """Throwing this exception results in 401 (Unauthorized) HTTP response code."""

    def __init__(self, message: str):
        """Inits unauthorized exception.

        Args:
            message: Exception message.
        """
        super().__init__(message, 401)


class ValidationException(ApiException):
    """Represents validation exception. Throwing this exception results in 400 (Bad Request) HTTP response code.

    Attributes:
        _details: Validation exception details
    """

    def __init__(self, message: str, details: List[dict] = None):
        """Inits validation error.

        Args:
            message: Exception message.
            details: Exception data.
        """
        super().__init__(message + ', check error.details for more information', 400)
        self._details = details

    @property
    def details(self) -> List[dict]:
        """Returns validation exception details.

        Returns:
            Validation exception details.
        """
        return self._details


class InternalException(ApiException):
    """Represents unexpected exception. Throwing this error results in 500 (Internal Error) HTTP response code."""

    def __init__(self, message: str):
        """Inits unexpected exception.

        Args:
            message: Exception message.
        """
        super().__init__(message, 500)


class TooManyRequestsErrorMetadata(TypedDict):
    periodInMinutes: float
    """Throttling period in minutes."""
    requestsPerPeriodAllowed: int
    """Available requests for periodInMinutes."""
    recommendedRetryTime: str
    """Recommended date to retry request."""


class TooManyRequestsException(ApiException):
    """Represents too many requests error. Throwing this exception results in 429 (Too Many Requests) HTTP response
    code."""

    def __init__(self, message: str, metadata: TooManyRequestsErrorMetadata):
        """Inits too many requests exception.

        Args:
            message: Exception message.
            metadata: Error metadata.
        """
        super().__init__(message, 429)
        self.metadata = metadata


examples/metastats/metastats.py
from metaapi_cloud_sdk import MetaStats, MetaApi
import asyncio
import os


async def example():
    # your MetaApi API token
    token = os.getenv('TOKEN') or '<put in your token here>'
    # your MetaApi account id
    account_id = os.getenv('ACCOUNT_ID') or '<put in your account id here>'
    # your MetaApi account
    account = None

    api = MetaApi(token)
    meta_stats = MetaStats(token)
    # you can configure http client via second parameter,
    # see in-code documentation for full definition of possible configuration options

    async def account_deploy():
        try:
            nonlocal account
            account = await api.metatrader_account_api.get_account(account_id)

            #  wait until account is deployed and connected to broker
            print('Deploying account')
            if account.state != 'DEPLOYED':
                await account.deploy()
            else:
                print('Account already deployed')
            print('Waiting for API server to connect to broker (may take couple of minutes)')
            if account.connection_status != 'CONNECTED':
                await account.wait_connected()

        except Exception as err:
            print(meta_stats.format_error(err))

    await account_deploy()

    async def get_account_metrics():
        try:
            metrics = await meta_stats.get_metrics(account_id)
            print(metrics)  # -> {'trades': ..., 'balance': ..., ...}
        except Exception as err:
            print(meta_stats.format_error(err))

    await get_account_metrics()

    async def get_account_trades(start_time, end_time):
        try:
            trades = await meta_stats.get_account_trades(account_id, start_time, end_time)
            print(trades[-5:])  # -> {_id: ..., gain: ..., ...}
        except Exception as err:
            print(meta_stats.format_error(err))

    await get_account_trades('2020-01-01 00:00:00.000', '2021-01-01 00:00:00.000')

    async def get_account_open_trades():
        try:
            trades = await meta_stats.get_account_open_trades(account_id)
            print(trades)  # -> {_id: ..., gain: ..., ...}
        except Exception as err:
            print(meta_stats.format_error(err))

    await get_account_open_trades()

asyncio.run(example())


lib/clients/metaStats_client.py
from .metaStats_models import Metrics, Trade, OpenTrade


class MetaStatsClient:
    """metaapi.cloud MetaStats MetaTrader API client"""
    def __init__(self, http_client, token: str, domain: str = None):
        """Inits MetaTrader API client instance.

        Args:
            http_client: HTTP client.
            token: Authorization token.
            domain: Domain to connect to, default is agiliumtrade.agiliumtrade.ai.
        """
        domain = domain or 'agiliumtrade.agiliumtrade.ai'
        self._httpClient = http_client
        self._host = f'https://metastats-api-v1.{domain}'
        self._token = token

    async def get_metrics(self, account_id: str, include_open_positions: bool = False) -> Metrics:
        """Returns metrics of MetaApi account. This API call is billable
        https://metaapi.cloud/docs/metastats/restApi/api/calculateMetrics/

        Args:
            account_id: MetaApi account id.
            include_open_positions: Indicates whether open positions will be included in the metrics, default false.

        Returns:
            Account metrics.
        """
        opts = {
            'url': f'{self._host}/users/current/accounts/{account_id}/metrics',
            'method': 'GET',
            'params': {
                'includeOpenPositions': include_open_positions
            },
            'headers': {
                'auth-token': self._token
            }
        }
        response = await self._httpClient.request(opts)
        return response['metrics']

    async def get_account_trades(self, account_id: str, start_time: str, end_time: str, update_history: bool = False,
                                 limit: int = 1000, offset: int = 0) -> Trade:
        """Returns historical trades of MetaApi account.
        https://metaapi.cloud/docs/metastats/restApi/api/getHistoricalTrades/

        Args:
            account_id: MetaApi account id.
            start_time: Start of time range, inclusive.
            end_time: End of time range, exclusive.
            update_history: Update historical trades before returning results. If set to true, the API call will be
            counted towards billable MetaStats API calls. If set to false, the API call is not billable.
            limit: Pagination limit.
            offset: Pagination offset.

        Returns:
            Account historical trades.
        """
        opts = {
            'url': f'{self._host}/users/current/accounts/{account_id}/historical-trades/{start_time}/{end_time}',
            'method': 'GET',
            'params': {
                'updateHistory': update_history,
                'limit': limit,
                'offset': offset
            },
            'headers': {
                'auth-token': self._token
            }
        }
        response = await self._httpClient.request(opts)
        return response['trades']

    async def get_account_open_trades(self, account_id: str) -> OpenTrade:
        """Returns open trades of MetaApi account. This API call is not billable.
        https://metaapi.cloud/docs/metastats/restApi/api/getOpenTrades/

        Args:
            account_id: MetaApi account id.

        Returns:
            Account historical trades.
        """
        opts = {
            'url': f'{self._host}/users/current/accounts/{account_id}/open-trades',
            'method': 'GET',
            'headers': {
                'auth-token': self._token
            }
        }
        response = await self._httpClient.request(opts)
        return response['openTrades']


lib/clients/httpClient.py
from .errorHandler import UnauthorizedException, ForbiddenException, ApiException, \
    ValidationException, InternalException, NotFoundException, TooManyRequestsException
from typing_extensions import TypedDict
from typing import Optional
from ..models import ExceptionMessage,  date
from .timeoutException import TimeoutException
import json
import asyncio
import sys
import httpx
from datetime import datetime
from httpx import HTTPError, Response


if sys.version_info[0] == 3 and sys.version_info[1] >= 8 and sys.platform.startswith('win'):
    asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())


class RetryOpts(TypedDict):
    retries: Optional[int]
    """Maximum amount of request retries, default value is 5."""
    minDelayInSeconds: Optional[float]
    """Minimum delay in seconds until request retry, default value is 1."""
    maxDelayInSeconds: Optional[float]
    """Maximum delay in seconds until request retry, default value is 30."""


class RequestOptions(TypedDict):
    """Options for HttpClient requests."""
    method: Optional[str]
    url: str
    headers: Optional[dict]
    params: Optional[dict]
    body: Optional[dict]
    files: Optional[dict]


class HttpClient:
    """HTTP client library based on requests module."""
    def __init__(self, timeout: float = None, retry_opts=None):
        """Inits HttpClient class instance.

        Args:
            timeout: Request timeout in seconds.
        """
        timeout = timeout or 60
        if retry_opts is None:
            retry_opts = {}
        self._timeout = timeout
        self._retries = retry_opts['retries'] if 'retries' in retry_opts else 5
        self._minRetryDelayInSeconds = retry_opts['minDelayInSeconds'] if 'minDelayInSeconds' in retry_opts else 1
        self._maxRetryDelayInSeconds = retry_opts['maxDelayInSeconds'] if 'maxDelayInSeconds' in retry_opts else 30

    async def request(self, options: RequestOptions, retry_counter: int = 0, end_time: float = None) -> Response:
        """Performs a request. Response errors are returned as ApiException or subclasses.

        Args:
            options: Request options.

        Returns:
            A request response.
        """
        if not end_time:
            end_time = datetime.now().timestamp() + self._maxRetryDelayInSeconds * self._retries
        retry_after_seconds = 0
        try:
            response = await self._make_request(options)
            response.raise_for_status()
            if response.status_code == 202:
                retry_after_seconds = response.headers['retry-after']
                if isinstance(retry_after_seconds, str):
                    retry_after_seconds = float(retry_after_seconds)
            if response.content:
                try:
                    response = response.json()
                except Exception as err:
                    print('Error parsing json', err)
        except HTTPError as err:
            retry_counter = await self._handle_error(err, retry_counter, end_time)
            return await self.request(options, retry_counter, end_time)
        if retry_after_seconds:
            await self._handle_retry(end_time, retry_after_seconds)
            response = await self.request(options, retry_counter, end_time)
        return response

    async def _make_request(self, options: RequestOptions) -> Response:
        async with httpx.AsyncClient(timeout=self._timeout) as client:
            method = options['method'] if ('method' in options) else 'GET'
            url = options['url']
            params = options['params'] if 'params' in options else None
            files = options['files'] if 'files' in options else None
            headers = options['headers'] if 'headers' in options else None
            body = options['body'] if 'body' in options else None
            req = client.build_request(method, url, params=params, files=files, headers=headers, json=body)
            response = await client.send(req)
            return response

    async def _handle_retry(self, end_time: float, retry_after: float):
        if end_time > datetime.now().timestamp() + retry_after:
            await asyncio.sleep(retry_after)
        else:
            raise TimeoutException('Timed out waiting for the end of the process of calculating metrics')

    async def _handle_error(self, err, retry_counter: int, end_time: float):
        if err.__class__.__name__ == 'ConnectTimeout':
            error = err
        else:
            error = self._convert_error(err)
        if error.__class__.__name__ in ['InternalException', 'ApiException'] and retry_counter < self._retries:
            pause = min(pow(2, retry_counter) * self._minRetryDelayInSeconds, self._maxRetryDelayInSeconds)
            await asyncio.sleep(pause)
            return retry_counter + 1
        elif error.__class__.__name__ == 'TooManyRequestsException':
            retry_time = date(error.metadata['recommendedRetryTime']).timestamp()
            if retry_time < end_time:
                await asyncio.sleep(retry_time - datetime.now().timestamp())
                return retry_counter
        raise error

    def _convert_error(self, err: HTTPError):
        try:
            response: ExceptionMessage or TypedDict = json.loads(err.response.text)
        except Exception:
            response = {}
        err_message = response['message'] if 'message' in response else err.response.reason_phrase
        status = err.response.status_code
        if status == 400:
            details = response['details'] if 'details' in response else []
            return ValidationException(err_message, details)
        elif status == 401:
            return UnauthorizedException(err_message)
        elif status == 403:
            return ForbiddenException(err_message)
        elif status == 404:
            return NotFoundException(err_message)
        elif status == 429:
            err_metadata = response['metadata'] if 'metadata' in response else {}
            return TooManyRequestsException(err_message, err_metadata)
        elif status == 500:
            return InternalException(err_message)
        else:
            return ApiException(err_message, status)
